api/app/utils/enphase_utils.py

```diff
@@ -0,0 +1,126 @@
+import http.client
+import pandas as pd
+from datetime import datetime, timedelta
+import os
+import xarray as xr
+import base64
+import json
+
+from quartz_solar_forecast.forecasts import forecast_v1_tilt_orientation
+from quartz_solar_forecast.pydantic_models import PVSite
+from quartz_solar_forecast.forecast import predict_tryolabs
+from quartz_solar_forecast.data import get_nwp, process_pv_data
+from quartz_solar_forecast.inverters.enphase import process_enphase_data
+
+def get_enphase_access_token(auth_code):
+    client_id = os.getenv("ENPHASE_CLIENT_ID")
+    client_secret = os.getenv("ENPHASE_CLIENT_SECRET")
+    redirect_uri = os.getenv("ENPHASE_REDIRECT_URI", "https://api.enphaseenergy.com/oauth/redirect_uri")
+
+    credentials = f"{client_id}:{client_secret}"
+    encoded_credentials = base64.b64encode(credentials.encode("utf-8")).decode("utf-8")
+
+    conn = http.client.HTTPSConnection("api.enphaseenergy.com")
+    headers = {
+        "Content-Type": "application/x-www-form-urlencoded",
+        "Authorization": f"Basic {encoded_credentials}"
+    }
+    payload = f"grant_type=authorization_code&redirect_uri={redirect_uri}&code={auth_code}"
+
+    conn.request("POST", "/oauth/token", payload, headers)
+    res = conn.getresponse()
+    data = res.read()
+    conn.close()
+
+    data_json = json.loads(data.decode("utf-8"))
+
+    if "error" in data_json:
+        raise ValueError(f"Error in getting access token: {data_json['error_description']}")
+
+    if "access_token" not in data_json:
+        raise KeyError(f"Access token not found in response. Response: {data_json}")
+
+    return data_json["access_token"]
+
+def get_enphase_data(enphase_system_id: str, access_token: str) -> pd.DataFrame:
+    api_key = os.getenv("ENPHASE_API_KEY")
+    start_at = int((datetime.now() - timedelta(weeks=1)).timestamp())
+    granularity = "week"
+
+    conn = http.client.HTTPSConnection("api.enphaseenergy.com")
+    headers = {"Authorization": f"Bearer {access_token}", "key": api_key}
+
+    url = f"/api/v4/systems/{enphase_system_id}/telemetry/production_micro?start_at={start_at}&granularity={granularity}"
+    conn.request("GET", url, headers=headers)
+    res = conn.getresponse()
+    data = res.read()
+    conn.close()
+
+    data_json = json.loads(data.decode("utf-8"))
+    return process_enphase_data(data_json, start_at)
+
+def make_pv_data(
```

Comment:
im confused here, why dont you use? https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/data.py#L234

Reply:
> im confused here, why dont you use? https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/data.py#L234

I'm passing the access token through run_forecast which uses make pv data and passing this token to every enphase request, but I think by using env vars as you suggested , this could be overcome

Reply:
did you do this with env vars?

Reply:
> did you do this with env vars?

Yes, within the enphase.py file, when an auth code is sent from the streamlit Frontend, it stores the access and refresh tokens in env vars